### ISSUE TITLE
Highlight the word result that searched

### ISSUE THREAD

**Issue by LIGMATV** [docsifyjs/docsify#2450]
#### Problem or desire

I searching for the "Beginilah" result

![Screenshot Capture - 2024-06-09 - 12-08-50](https://github.com/docsifyjs/docsify/assets/143163098/b0251ae2-47f2-4385-93e7-ef07e8c05d63)

When i click the result, it just go to the Heading of text, not highlight the word.

#### Proposal

Highlight the word that searched, so we know where it is

![Screenshot Capture - 2024-06-09 - 12-10-06](https://github.com/docsifyjs/docsify/assets/143163098/3820618a-b30d-4b1e-a0ff-dd6449e65369)

#### Implementation

Add new element for the highlited searched words like ``<mark class="searched"``...

**Comment by Koooooo-7** [comment]
Hi @LIGMATV , TBH, it is not that feasible to docsify since the search function (plugin) and the main contents are isolated.
It is unlike the search engine stuff that docsify does not do the global text search and highlights base on all the contents files before render it to HTML. In docsify, it is not correlative.
The search function works on its own INDEXs and when you click the search result to tell docsify `where the anchor I want to go`, it triggers main content to load.

**Comment by jhildenbiddle** [comment]
I can think of a few different ways to handle this fairly easily.

**Option 1:** The search plugin can add a separate plugin to the plugins array that determines if a search is currently active and, if so, parses and marks content before it is rendered. 

**Option 2:** When navigation occurs, the `data-page` attribute on the `body` is updated. Currently, this update occurs when a link is clicked and not when a page has loaded. If we change the behavior to update `data-page` after loading is complete, the search plugin could use a mutation observer on the `<body>` element, listening *only* for changes to this specific property. When a change is detected, the DOM can be walked and elements marked as needed.

The most challenging part will likely be determining which matches to mark and which ones to ignore. For example, we don't want mark matches within a `<script>`, `<textarea>`, `<option>`, comments, etc.

**Comment by Koooooo-7** [comment]
Hi @jhildenbiddle, I think `highlight` search content in main content is not that meaningful and maintain on our side.

It is not a normal behavior for search engines or for a documentation site. The 'highlight' feature is used to display the matched contents in the search result list, not in the main content. This holds true even when using `ElasticSearch` or other search implementations in e-commerce search functionalities.

And back to the docs site functions itself:
1. highlighting incomplete search results in the text can also disrupt the reading experience.
2. single search keyword is meaningless in a site without any context. Unless you whole page/sites is a pure `configurations` docs without any other contents for the configurations list.
 
Additionally, the browser search function for the search matched target page is more convenient if really needs.

**Comment by jhildenbiddle** [comment]
@Koooooo-7 --

I don't think there's a right or wrong answer on this one. It's a preference that can change based on the user, the type of search being performed, or where a matching search term is located in the content area. 

My comment above was focused more on how we could implement highlighting matching search terms in the content area in response to the "it is not that feasible" statement in the comment prior to mine. Regarding if we should offer this feature, here are a few things to consider...

- A quick review of Docsify competitors like Docusaurus, GitBook, Starlight, and VitePress shows that--to your point--none of them highlight matching search terms in the content area when a search is active. That said, we should consider these projects but not allow their features or limitations to dictate our own roadmap. Perhaps they do not offer this feature *yet* but will in the future, or perhaps they chose not to prioritize it because the third-party search tool they are using makes implementation more challenging (Starlight uses [Pagefind](https://pagefind.app), Docusaurus uses [Algolia](https://www.algolia.com/), VitePress uses [minisearch](https://lucaong.github.io/minisearch/), etc.). The point is that we shouldn't assume they made a conscience decision not to implement this feature unless we have reason to believe otherwise.
- Like Docsify, all of the competitors above link to a heading rather than the matching search term in the content. There are pros and cons to this: the heading provides context for the matching search term it contains (pro), but the matching search term can be hard to find when it is part of a large block of content under a heading (con). The ability to quickly scan for a highlighted search term is, I believe, the primary reason why someone would want to highlight matching search terms in the content area.
- The browser's built-in page search functionality can be used to find matching search terms in the content area, but it has issues: 1) it requires typing in the search term a second time and 2) the search begins at the top of the page and not the content area which means it will match all of the results in the sidebar before the content area.

> highlighting incomplete search results in the text can disrupt the reading experience.

Not sure what you mean by "incomplete" search results. The behavior I am referring to is highlighting matching search results on pages loaded by clicking on a search result in a sidebar. When an active search term is modified, the previous search results are cleared as well as any highlighted content in the content area. We would not do "live" highlighting of the content area based on what a user types in our search field.

If users find content highlighting distracting, users can disable the feature (assuming a user-facing toggle is provided) or clear the search results via ESC or the clear button. The content and page position will remain unchanged when this occurs.

> single search keyword is meaningless in a doc without any context

There is no way to know if this statement is accurate for every site or search. We cannot assume a heading is context needed for all matching search terms. It is equally likely that the text surrounding the matching search term provides more/better context than the heading the search term resides under, especially if the search term and surrounding text are part of a large block of content, a callout/admonishment, a list item, part of a code block, etc.

> Additionally, the browser search function for the search matched target page is more convenient if really needs.

See above.

---

My preference regarding this feature in order of preference:

1. Provide a small "Highlight content" (or similar text) toggle below the search field when a search results is active. This allows site visitors to decide if they want matching content to be highlighted or not.
2. Offer this as a search plugin option (e.g., `highlightContent`). This allows site maintainers to decide if they want matching content to be highlighted or not. Ideally, a user-facing toggle would still be displayed and setting this option would simply set the default state of the toggle so site visitors still retain control over this feature.
3. Wait for someone else to implement this feature as a separate plugin which could.

Personally, I like option 1 with the toggle set to "off" by default.

**Comment by Koooooo-7** [comment]
@jhildenbiddle  
I don't mean there is a `right or wrong` thing, but based on the product itself, the requirements have the priority and the `unnecessary` stuff, based on the user experiences, project perspective, the functionality persona, and the `conventions` (don't mean something that we follow the competitors).

If the user requirements is not a *common use case nor the common functions* cross the competitors or other apps, and it is not a brand new things only we can do, I don't think we need bring it in. 

> A quick review of Docsify competitors like Docusaurus, GitBook, Starlight, and VitePress shows that--to your point--none of them highlight matching search terms in the content area when a search is active. That said, we should consider these projects but not allow their features or limitations to dictate our own roadmap. Perhaps they do not offer this feature yet but will in the future, or perhaps they chose not to prioritize it because the third-party search tool they are using makes implementation more challenging (Starlight uses [Pagefind](https://pagefind.app/), Docusaurus uses [Algolia](https://www.algolia.com/), VitePress uses [minisearch](https://lucaong.github.io/minisearch/), etc.). The point is that we shouldn't assume they made a conscience decision not to implement this feature unless we have reason to believe otherwise.

I'm not a follower that we only do others do, but again, I consider about the product itself.
Besides, the google, wiki,  and other normal websites/apps, they don‘t have this functions either.
Although they implement the search functions all by themselves with [`ElasticSearch`](https://www.elastic.co/elasticsearch) I mentioned or other advanced search solutions.

I participated design and know few （3? IIUC） of sites do have the `contents` highlight functions as an `advanced search` function also.
(They have a label/button on the page top (such as `Highlighting keyword "XXX"`) as a toggle to disable the highlight like your idea.) And they are almost some more complex mixed media type tools site.

> We cannot assume a heading is the only context needed/wanted for all matching search terms on every site. 

Yes, but it brings the thing what we normally need:
This contents blowing this heading is related to the search things.

>  It is equally likely that the text surrounding a matching search term provides more useful context than the heading the search term resides under, in which case allowing users acquire this context as quickly as possible should be the goal.

Yes, I agree, and this is what makes this requirement reasonable. 

> This is often true for matching search terms found in a large blocks of content, callouts/admonishments, list items, code blocks, etc.

For a short page/section, they highlight or not doesn't matter.

For a long content, such as highlight a function name `foo` inside a 5000 lines `code block`.  It only useful when the search content is `unique` as I said. Otherwise, the searched keyword has duplicated in the target page/ code block.
What should user do?
- Click another search matched keyword result (maybe wrong again, things the search result brings limited context, unlike the SEO stuff brings to the spider).
- Use the browser result to fast jump to next search result. (If the highlight keywords can not brings this function, the highlight is also meaningless)

Personally, normally, I don't think user will scroll down to find the highlight part in the 5000 lines code block.

> highlighting incomplete search results in the text can disrupt the reading experience.

highlighting incomplete search means I input the `docum `(entation), then I click the search target page, it highlight this `docum` is also annoying. Although we could say `Although we can say that this highlighted area is a sign`.
It makes the search result has higher rate to duplicated also.

---

I believe that we can make it.
I only consider should we brings it to our side as a common feature.
If some implement it as a plugin or they customize docsify by themselves is fine to me.

**Comment by jhildenbiddle** [comment]
Agreed on the priority. This doesn't need to be part of v5 release, but it could be handled in a v5.x point release in the future. I'm fine leaving I tagged as "Future" for now.
